### DISCUSSION TITLE
ld: Undefined symbols

### DISCUSSION THREAD

**Discussion by Arbifoxx** [ChefKissInc/Inferno#16] [General]
Hello! I am trying to compile this, but I keep running into this error. Any help would be appreciated. Also, as a side note, what's the status on springboard? I've noticed that recent commits have referenced previous issues so I'm wondering if progress has been made on SEP. Thank you!

ld: warning: -undefined error is deprecated
ld: warning: ignoring duplicate libraries: '-lgcrypt', '-lgpg-error'
ld: Undefined symbols:
  _asn1_array2tree, referenced from:
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
  _asn1_create_element, referenced from:
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
  _asn1_der_decoding, referenced from:
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
  _asn1_read_value, referenced from:
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
      _extract_im4p_payload in hw_arm_apple-silicon_boot.c.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
make: *** [run-ninja] Error 1

**Comment by VisualEhrmanntraut** [comment]
@waterdragon78 https://github.com/TrungNguyen1909/qemu-t8030/wiki/Bringing-up-the-emulator#getting-dependencies

**Comment by Arbifoxx** [comment]
I did that, and now I'm getting this
ld: warning: -undefined error is deprecated
ld: warning: ignoring duplicate libraries: '-lc++', '-lgcrypt', '-lglib-2.0', '-lgmodule-2.0', '-lgpg-error', '-lintl'
ld: Undefined symbols:
  _asn1_array2tree, referenced from:
      _extract_im4p_payload in hw_arm_xnu.c.o
  _asn1_create_element, referenced from:
      _extract_im4p_payload in hw_arm_xnu.c.o
  _asn1_der_decoding, referenced from:
      _extract_im4p_payload in hw_arm_xnu.c.o
  _asn1_read_value, referenced from:
      _extract_im4p_payload in hw_arm_xnu.c.o
      _extract_im4p_payload in hw_arm_xnu.c.o
      _extract_im4p_payload in hw_arm_xnu.c.o
      _extract_im4p_payload in hw_arm_xnu.c.o
      _extract_im4p_payload in hw_arm_xnu.c.o
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[3081/3116] Compiling C object tests/qtest/boot-order-test.p/boot-order-test.c.o
[3082/3116] Compiling C object tests/qtest/erst-test.p/erst-test.c.o
[3083/3116] Compiling C object tests/qtest/fw_cfg-test.p/fw_cfg-test.c.o
[3084/3116] Compiling C object tests/qtest/device-plug-test.p/device-plug-test.c.o
[3085/3116] Compiling C object tests/qtest/fdc-test.p/fdc-test.c.o
[3086/3116] Compiling C object tests/qtest/i440fx-test.p/i440fx-test.c.o
[3087/3116] Compiling C object tests/qtest/drive_del-test.p/drive_del-test.c.o
[3088/3116] Compiling C object tests/qtest/cpu-plug-test.p/cpu-plug-test.c.o
[3089/3116] Compiling C object tests/qtest/ide-test.p/ide-test.c.o
[3090/3116] Compiling C object tests/qtest/hd-geo-test.p/hd-geo-test.c.o
[3091/3116] Compiling C object tests/qtest/rtc-test.p/rtc-test.c.o
[3092/3116] Compiling C object tests/qtest/q35-test.p/q35-test.c.o
[3093/3116] Compiling C object tests/qtest/tco-test.p/tco-test.c.o
ninja: build stopped: subcommand failed.
make: *** [run-ninja] Error 1

**Comment by VisualEhrmanntraut** [comment]
@waterdragon78 You did not install asn1

**Comment by Arbifoxx** [comment]
I did, using venv and pip

**Comment by VisualEhrmanntraut** [comment]
@waterdragon78 How’s it gonna access the venv

**Comment by Arbifoxx** [comment]
It wouldn't let me install it otherwise
![Screenshot 2024-02-27 at 11 04 27 AM](https://github.com/ChefKissInc/qemu-apple-silicon/assets/45146536/c9155d0d-f691-408d-a4a7-62cebbdfd90e)

**Comment by Arbifoxx** [comment]
I used the flag, but I still have the same error

**Comment by VisualEhrmanntraut** [comment]
pyasn1 doesn't have anything to do with the asn1 lib here

**Comment by Arbifoxx** [comment]
I compiled it with Ubuntu which worked just fine, but I'm trying to run the command below to start it but it can't find a property.

I used:

qemu-t8030/build/qemu-system-aarch64 -s -M t8030,trustcache-filename=Firmware/038-44135-124.dmg.trustcache,ticket-filename=root_ticket.der -kernel kernelcache.research.iphone12b -dtb Firmware/all_flash/DeviceTree.n104ap.im4p -append "debug=0x14e kextlog=0xffff serial=3 -v wdt=-1" -initrd 038-44135-124.dmg -cpu max -smp 4 -m 4G -serial mon:stdio -drive file=nvme.1,format=raw,if=none,id=drive.1 -device nvme-ns,drive=drive.1,bus=nvme-bus.0,nsid=1,nstype=1,logical_block_size=4096,physical_block_size=4096 -drive file=nvme.2,format=raw,if=none,id=drive.2 -device nvme-ns,drive=drive.2,bus=nvme-bus.0,nsid=2,nstype=2,logical_block_size=4096,physical_block_size=4096 -drive file=nvme.3,format=raw,if=none,id=drive.3 -device nvme-ns,drive=drive.3,bus=nvme-bus.0,nsid=3,nstype=3,logical_block_size=4096,physical_block_size=4096 -drive file=nvme.4,format=raw,if=none,id=drive.4 -device nvme-ns,drive=drive.4,bus=nvme-bus.0,nsid=4,nstype=4,logical_block_size=4096,physical_block_size=4096 -drive file=nvram,if=none,format=raw,id=nvram -device apple-nvram,drive=nvram,bus=nvme-bus.0,nsid=5,nstype=5,id=nvram,logical_block_size=4096,physical_block_size=4096 -drive file=nvme.6,format=raw,if=none,id=drive.6 -device nvme-ns,drive=drive.6,bus=nvme-bus.0,nsid=6,nstype=6,logical_block_size=4096,physical_block_size=4096 -drive file=nvme.7,format=raw,if=none,id=drive.7 -device nvme-ns,drive=drive.7,bus=nvme-bus.0,nsid=7,nstype=8,logical_block_size=4096,physical_block_size=4096 -monitor telnet:127.0.0.1:1235,server,nowait
(I renamed it to qemu-t8030 for ease of access, its the dev-sw-sep branch)

Which gives:

Property 't8030-machine.trustcache-filename' not found

**Comment by VisualEhrmanntraut** [comment]
@waterdragon78 Why are you using the dev branch

**Comment by Arbifoxx** [comment]
Because it has more commits and therefore progress

**Comment by VisualEhrmanntraut** [comment]
but it doesn't restore because it's a W.I.P so what's the point

**Comment by Arbifoxx** [comment]
Where does it state that?

**Comment by VisualEhrmanntraut** [comment]
and where does it state the opposite

**Comment by Arbifoxx** [comment]
That’s a valid point, mb. Do both dev branches not work?

**Comment by VisualEhrmanntraut** [comment]
my local changes on the SW SEP branch reach 92% on restore but fails on update_sep because of some weird issue :(
```
entering update_sep
libMobileGestalt MobileGestalt.c:711: RestoreLog: Client Query: Image4CryptoHashMethod
2024-03-01 20:47:14.469213+0000 restored_external[8:433] RestoreLog: Client Query: Image4CryptoHashMethod
libMobileGestalt MobileGestalt.c:743: RestoreLog: Client Response: Image4CryptoHashMethod : sha2-384
2024-03-01 20:47:14.473331+0000 restored_external[8:433] RestoreLog: Client Response: Image4CryptoHashMethod : sha2-384
ramrod_copy_manifest_digest_data_from_img4: failed to parse img4Data as Img4
Fri Mar  1 20:47:14 2024 localhost com.apple.xpc.launchd[1] <Error>: Host-level exception raised: pid = 8, thread = 0xe33, exception type = 0xd, codes = { 8589934603 }, states = { 0 }
[20:47:14.0506-GMT]{1>4} CHECKPOINT NOTICE: Post-engines NVRAM variable: restore-outcome=initial_engine_no_return
[20:47:14.0539-GMT]{1>4} CHECKPOINT PROGRESS: NO_REBOOT_RETRY (initial_monitor_no_return) -> (recovery_mode)
restore-child-failures = {initial:SIGNAL(11)[REBOOT]}
executing /usr/sbin/nvram restore-child-failures={initial:SIGNAL(11)[REBOOT]}
restore-outcome = recovery_mode
executing /usr/sbin/nvram restore-outcome=recovery_mode
restored exited with status 0xb due to signal 11 - rebooting
```

**Comment by Arbifoxx** [comment]
I’m not an expert but I think it’s failing some sort of hash and then failing to parse img4data as img4? I’m not sure what that is lol but maybe it’s expecting it to be unchanged if it is changed.

**Comment by VisualEhrmanntraut** [comment]
I think idevicerestore is sending something wrong but idk the code is kind of unreadable lol

**Comment by Arbifoxx** [comment]
Idk assuming I can fix the unknown property issue and get it running I’ll try it on Linux. Maybe that’ll do something assuming yall are on Mac? Idk, maybe something’s different in the Linux idevicerestore. Slim chance but can’t hurt to test it. Just gotta get it running first tho lol.

**Comment by Arbifoxx** [comment]
I still have no idea what's going on, is there any way to fix the unknown property?

**Comment by VisualEhrmanntraut** [comment]
@waterdragon78 remove `-filename` from every property

**Reply by Arbifoxx** [reply]
thx

**Reply by Arbifoxx** [reply]
I got it working, and I wanted to try the dev branch. Now it's telling me that I need an SEPROM, what arguement and file should I get for that?

**Reply by VisualEhrmanntraut** [reply]
You need to find a decrypted T8020 SEPROM and SEPFW. You have to figure out the rest, Google is your friend.

**Reply by Arbifoxx** [reply]
Gotcha, I can do that. Whats the arguement to use for that?

**Reply by VisualEhrmanntraut** [reply]
... `seprom` and `sepfw`

**Reply by Arbifoxx** [reply]
I only did the seprom because that's all I could find but I still got the apple logo and restore logo. Would that still boot?

**Reply by VisualEhrmanntraut** [reply]
As I said previously, dev will NOT boot; restore will fail prematurely.
